### ISSUE TITLE
Access to parent context

### ISSUE THREAD

**Issue by guffy1234** [FluentValidation/FluentValidation#1942]
### Is your feature request related to a problem? Please describe.

Hi
There is issue #904  that describes that the internal property PassThroughParentContext somehow (?) could be set to true to pass through parent context to children.
Why it still internal and how set it?
Why it is a bad idea to pass parent context all the time?

### Describe the solution you'd like

Remove this property and pass parent every time.

### Describe alternatives you've considered

Add yet another functor signature that will allow to give value for PassThroughParentContext

**Comment by JeremySkinner** [comment]
The issue you linked to is from 2018 and no longer relevant - the property in question is no longer used.

**Comment by guffy1234** [comment]
Hi @JeremySkinner 
I thinking about how to pass arbitrary data to validators via some context.
I found this example, 
https://gist.github.com/maca88/9e4551da22cf53056143bef76148a452
but quickly realized that it is based on pretty old FluentValidator library.
After that I found that there is dictionary RootContextData that mostly solves my needs. But the idea to use custom root validation context with typed properties have not leave me and I was trying to implement something like this with traversing by using ParentContext property. Then I studied your codes and found that it pretty difficult, almost impossible due to there is no legal way to enable ParentContext assignment. 
As for me I don't see any reason why it not just propagated by default. So I just tried to move your attention to this small issue. Mayby it's a time to change something.

**Comment by JeremySkinner** [comment]
So in modern versions the `context` parameter is the same as `context.ParentContext` from older versions. 

For some background, older versions of FluentValidation (as per the example you linked to) received a `PropertyValidatorContext` which was a new/separate instance for each rule. ParentContext then allowed you to access the parent `ValidationContext<T>`. Modern versions of FluentValidation don’t work this way - PropertyValidatorContext no longer exists, you instead receive the `ValidationContext<T>` directly (which is a shared instance across all rules; this was done for performance reasons). 

Contexts are no longer created for each rule but are instead shared (for performance), and there is no longer a parent. New ones are now only created when using child validators (as the type parameters are different so the same instance can’t be reused). 

All that being said, RootContextData is the correct way of doing what you want.

**Comment by guffy1234** [comment]
I was so quick so even created a pull request )
I think that the link between child context and parent from which it had been forked is useful. IMHO. Or all useless codes shall be removed just because they only confuse. 
But it's up to you of course. You may ignore the PR.

Btw, it would be nice to hint a user to use Object Pool to deal with validators. Because constructing a validator inspires a lot of allocations. A user of the library should be aware that he can reuse the entire constructed validator for subsequent requests.
Or the library can even provide some support of such pools out of the box.
https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.objectpool.defaultobjectpool-1?view=dotnet-plat-ext-6.0

**Comment by JeremySkinner** [comment]
The PR will fail the build at the moment as it introduces a breaking change to the public api. You’d need to introduce a new overload rather than changing the method signature. If you can fix that then I’ll review it, but I can’t promise I’ll merge and I’ll need to take some time to think through the implications. 

> Btw, it would be nice to hint a user to use Object Pool to deal with validators

We won’t be doing this out of the box. This is the job of a DI container to handle validator lifetime. For web apps we generall recommend using the Scoped lifetime in. Although a singleton lifetime could technically perform better, this is often a source of confusion for users who aren’t aware of the complexities of working with singleton scoped objects (particularly if you try and make a singleton scoped object depend on something that’s not singleton scoped). We provide the FluentValidation.DependencyInjectionExtensions package that automates registering validators with a container where you can specify the lifetime.

**Comment by guffy1234** [comment]
I fixed the PR and reverted changes in public API.

You see, the documentation doesn't say that a validator instance is thread-safe.  
At first look, the https://github.com/FluentValidation/FluentValidation/blob/main/src/FluentValidation/AbstractValidator.cs
is looked like this. And a user may register a validator as a singleton.
But the user must ensure that there is no edge case when someone will try to all a rule on the fly via some public method.
Because enumeration of a collection is not thread safe. 
In other words, in a trivial case when rules are constructed by the constructor (as usual) singleton approach will work. Until someone somewhy will not write some "rules workaround" and introduce the hadly detectable bug.
The pool approach is much safer.

**Comment by JeremySkinner** [comment]
Thanks for making the update, once you’ve signed the CLA I’ll be able to progress the PR further.
